version/git.go

```diff
@@ -0,0 +1,231 @@
+package main
+
+import (
+	"context"
+	"dagger/version/internal/dagger"
+	"slices"
+	"strings"
+
+	"golang.org/x/mod/semver"
+)
+
+func (v *Version) Git() *Git {
+	ctr := dag.Wolfi().
+		Container(dagger.WolfiContainerOpts{Packages: []string{"git"}}).
+		WithWorkdir("/src")
+	if v.Inputs != nil {
+		ctr = ctr.WithDirectory(".", v.Inputs)
+	}
+	if v.GitDir != nil {
+		ctr = ctr.WithDirectory(".", v.GitDir)
+
+		// do various unshallowing operations (only the bare minimum is
+		// provided by the core git functions which are used by our remote git
+		// module sources)
+		remote := "https://github.com/dagger/dagger.git"
+		ctr = ctr.
+			// we need the unshallowed HEAD, so we can determine which tags are
+			// in HEAD's history later
+			WithExec([]string{"git", "fetch", "--unshallow", remote}).
+			// we need all the tags to determine the latest tags later
+			WithExec([]string{"git", "fetch", "--no-tags", remote, "+refs/tags/*:refs/tags/*"}).
+			// we need main, so we can determine the merge base for it later
+			WithExec([]string{"git", "fetch", "--no-tags", "-u", remote, "refs/heads/main:refs/heads/main"})
```

Comment:
It *really* sucks that we need all this extra `fetch`ing here, but we can do some really really cool things with it (e.g. we work out the last common `main` commit, so we are sure that that tag will actually exist).

However - contextual git should be able to handle this one day, we should be able to allow "unshallowed" clones.

Reply:
Note to self: maybe we could do these steps conditionally, by detecting if the repo is a shallow clone first (that way we don't need it for the local case) - that way we can avoid doing network access if we don't need to.

Reply:
As discussed in discord: it's totally fine for now. We can optimize later.